tests/caffe2_ref_test.py

```diff
@@ -27,6 +27,109 @@
 from caffe2.python.models.download import downloadFromURLToFile, getURLFromName, deleteDirectory
 
 
+class NodeSpec(object):
+    """
+    Describes a onnx.NodeProto, but without inputs/outputs
+    (which will be inferred).
+    """
+    def __init__(self, name, **kwargs):
+        self.name = name
+        self.kwargs = kwargs
+
+
+N = NodeSpec
+
+
+def create_input(call_args):
+    """
+    Creates a tuple of Numpy ndarray from a template 'call_args'; every embedded
+    tuple in call_args is converted into a random ndarray with the specified
+    dimensions.
+    """
+    if not isinstance(call_args, tuple):
+        call_args = (call_args,)
+
+    def map_arg(arg):
+        if isinstance(arg, tuple):
+            return np.random.randn(*arg).astype(np.float32)
+        else:
+            return arg
+
+    return tuple(map_arg(arg) for arg in call_args)
+
+
+class Caffe2NumpySpec(object):
+    """
+    A test for a single ONNX node against a Numpy reference implementation.
+    """
+    def __init__(self, name, node, np_impl, inputs):
+        """
+        Arguments:
+            name (string): Eventual name of the test.  Must be prefixed
+                with 'test_'.
+            node (NodeSpec): The ONNX node's name and attributes to be tested;
+                inputs and outputs will be inferred from other arguments of this
+                spec.
+            np_impl (lambda): A function from any number of Numpy ndarrays,
+                to a single ndarray or tuple of ndarrays.
+            inputs (tuple of ndarrays or size tuples): A specification of
+                the input to the operator.
+        """
+        # We don't prepend the 'test_' prefix to improve greppability
+        if not name.startswith("test_"):
+            raise ValueError("Test name must start with test_")
+        self.name = name
+        self.node = node
+        self.np_impl = np_impl
+        self.inputs = inputs
+
+    def run(self, test_self):
+        # TODO: In some cases we should generate multiple random inputs
+        # and test (ala Hypothesis)
+        args = create_input(self.inputs)
+        np_results = self.np_impl(*args)
+        if not isinstance(np_results, tuple):
+            np_results = (np_results, )
+        input_names = list(map(lambda i: "input_{}".format(i), range(len(args))))
```

Comment:
nit: use list comprehension? that way you don't need to worry whether map returning a list or a generator.